convert.go

```diff
@@ -0,0 +1,147 @@
+// Package flatfs is a Datastore implementation that stores all
+// objects in a two-level directory structure in the local file
+// system, regardless of the hierarchy of the keys.
+package flatfs
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+)
+
+func Convert(oldPath string, getOldDir ShardFunc, newPath string, getNewDir ShardFunc, out io.Writer) error {
+	old, err := New(oldPath, getOldDir, false)
+	if err != nil {
+		return err
+	}
+	new, err := New(newPath, getNewDir, false)
+	if err != nil {
+		return err
+	}
+
+	if out != nil {
+		fmt.Fprintf(out, "Getting Keys...\n")
+	}
+	res, err := old.Query(query.Query{KeysOnly: true})
+	if err != nil {
+		return err
+	}
+	entries, err := res.Rest()
+	if err != nil {
+		return err
+	}
+	prog := newProgress(len(entries), out)
+
+	if out != nil {
+		fmt.Fprintf(out, "Moving Keys...\n")
+	}
+
+	// first move the keys
+	for _, e := range entries {
+		err := moveKey(old, new, datastore.RawKey(e.Key))
+		if err != nil {
+			return err
+		}
+		prog.Next()
+	}
+
+	if out != nil {
+		fmt.Fprintf(out, "\nCleaning Up...\n")
+	}
+
+	// now walk the old top-level directory
+	dir, err := os.Open(old.path)
+	if err != nil {
+		return err
+	}
+	defer dir.Close()
+	names, err := dir.Readdirnames(-1)
+	if err != nil {
+		return err
+	}
+	for _, fn := range names {
+		if fn == "." || fn == ".." {
+			continue
+		}
+		oldPath := filepath.Join(old.path, fn)
+		inf, err := os.Stat(oldPath)
+		if err != nil {
+			return err
+		}
+		if inf.IsDir() {
+			// if we are a directory we should be empty so just remove it
+			err := os.Remove(oldPath)
+			if err != nil {
+				return err
+			}
+		} else {
+			// else move it
+			newPath := filepath.Join(new.path, fn)
+			err := os.Rename(oldPath, newPath)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if out != nil {
+		fmt.Fprintf(out, "All Done.\n")
+	}
+
+	return nil
+}
+
+func moveKey(old *Datastore, new *Datastore, key datastore.Key) error {
+	_, oldPath := old.encode(key)
+	dir, newPath := new.encode(key)
+	err := new.makeDirNoSync(dir)
+	if err != nil {
+		return err
+	}
+	err = os.Rename(oldPath, newPath)
```

Comment:
The idea is that source and destination should be on the same filesystem.   I was assuming this will fail on unix/OS X also.  The conversion operation will take a lot longer if a copy has to be made.